docs/rfcs/14-gdrive-for-go-storage-design.md

```diff
@@ -0,0 +1,31 @@
+- Author: Jun jun@junz.org
+- Start Date: 2021-7-18
+- RFC PR: [beyondstorage/go-service-gdrive#14](https://github.com/beyondstorage/go-service-gdrive/issues/14)
+- Tracking Issue: [beyondstorage/go-service-gdrive#15](https://github.com/beyondstorage/go-service-gdrive/issues/15)
+
+# GSP-14: Gdrive for go-storage design
+
+## Background
+
+Google drive API has so many different notions that differs from `go-storage`, and we have briefly discussed in [Gdrive use FileId to manipulate data instead of file name #11](https://github.com/beyondstorage/go-service-gdrive/issues/11). Now I would like to start a RFC so that we can make all things more clear.
+
+In Google drive API, `FileID` is a critical attribute of a file(or directory). We will use it to manipulate data instead of by path. In fact, path is very trivial in gdrive, and we can create files with the same name in the same location. In other words, path can be duplicate in gdrive. This behavior can cause some problems to our path based API.
+
+## Proposal
+
+**We manually stipulate that every path is unique.** When users try to call `Write` to an existing file, we update it's content instead of creating another file with the same name.
+
+**We will do a conversion between path and `FileID`.** In this way, every path can be converted to `FileID`, so we are able to build a good bridge between `go-storage` API and gdrive API.
+
+## Implementation
+
+When users try to call `Write("foo/bar/test.txt")`, first we use `pathToId` to try to convert the path to `FileID`. If it fails, then it means that this file doesn't exist, and we should create one. Then we can create folder `foo` , `bar`, and last the file `test.txt`. 
+
+Our significant point `pathToId` can be implement like this:
+
+If the file is in the root folder, then we just do a simple search by using `drive.service.Files.List().Q(searchArgs).Do()`. The return value type is `*drive.File`, and it's attribute `ID` is what we need.
+
+But if the file path is like `foo/bar/demo.txt`, it would be a little complex.
```

Comment:
If the path is complex like `a/b/c/d/e/f/h/q.txt`, it looks like we need to repeat the search many times.

How about cache the `path -> id` in memory with TTL?

Reply:
ping @junaire for a look.

Reply:
OK, I'll work on it. However, I don't have too much experience with this and I would like to ask if we should use some third library? Or just simply use the map from the standard library?

Reply:
To simplify the implementation, use a map with Mutex is OK for now.

We can implement the TTL logic later by https://github.com/dgraph-io/ristretto or other libs.